### PR TITLE
[EraVM] Optimize overflow intrinsic which has a branching use

### PR DESCRIPTION
**Pull request by TerryGuo**
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EraVM/EraVMISelLowering.cpp [hunk 1]
```diff
@@ -1105,6 +1105,41 @@ SDValue EraVMTargetLowering::LowerSELECT(SDValue Op, SelectionDAG &DAG) const {
...
+SDValue EraVMTargetLowering::LowerBRCOND(SDValue Op, SelectionDAG &DAG) const {
+  SDValue Chain = Op.getOperand(0);
+  SDValue Cond = Op.getOperand(1);
+  SDValue Dest = Op.getOperand(2);
+  SDLoc DL(Op);
+  SDValue MatchedUArithO = matchingOverflowArithmeticOperation(Cond);
+  if (!MatchedUArithO)
+    return SDValue();
+
+  // Similar with the approach in function LowerSELECT, we will end up with two
```
Comment (akiramenai) [inline llvm/lib/Target/EraVM/EraVMISelLowering.cpp]: Nit: I wouldn't reference another function. It's not needed to understand what is happening while comments imo to reduce number of necessary lookups needed for understanding.

File: llvm/lib/Target/EraVM/EraVMPostCodegenPrepare.cpp [hunk 2]
```diff
@@ -120,6 +135,97 @@ bool EraVMPostCodegenPrepare::runOnFunction(Function &F) {
...
+        // check extractvalue: there must be at least one use which is
+        // extractvalue and index 1
+        auto *ExtractValue = dyn_cast<ExtractValueInst>(U);
+        if (!ExtractValue ||
+            (!ExtractValue->hasIndices() || ExtractValue->getIndices()[0] != 1))
+          continue;
+
+        // check that the extracted value is used by a conditional branch in the
+        // same basicblock:
+        auto it = std::find_if(
```
Comment (akiramenai) [inline llvm/lib/Target/EraVM/EraVMPostCodegenPrepare.cpp]: nit: can we use `llvm::` counterpart?

**Review by akiramenai — COMMENTED** [review]
We are changing the asm syntax. To not interfere, I suggest to postpone merging it. But otherwise the patch looks good to me.

File: llvm/lib/Target/EraVM/EraVMISelLowering.cpp (hunk 1 above)
Comment (TerryGuo) [reply llvm/lib/Target/EraVM/EraVMISelLowering.cpp]: Updated now.

File: llvm/lib/Target/EraVM/EraVMPostCodegenPrepare.cpp (hunk 2 above)
Comment (TerryGuo) [reply llvm/lib/Target/EraVM/EraVMPostCodegenPrepare.cpp]: Updated now.

**Comment by vladimirradosavljevic** [comment]
I think this is a good PR to do some experiments to hopefully increase performance further:
1. Implement `shouldFormOverflowOp` in `EraVMTargetLowering`. I suggest first implementing it like this:
```
  bool shouldFormOverflowOp(unsigned Opcode, EVT VT,
                            bool MathUsed) const override {
    // Form add and sub with overflow intrinsics regardless of any extra
    // users of the math result.
    return VT == MVT::i256;
  }
```
And after that try with this, to compare the numbers:
```
  bool shouldFormOverflowOp(unsigned Opcode, EVT VT,
                            bool MathUsed) const override {
    return MathUsed && VT == MVT::i256;
  }
```
2. Remove [this](https://github.com/matter-labs/era-compiler-llvm/blob/aeeec90f885ddde68fab1fefb658431f8c579fe3/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L4920) constraint, since we now support umulo.

@TerryGuo @akiramenai What do you think about this?

**Comment by TerryGuo** [comment]
Hi @vladimirradosavljevic , thanks for your suggestions and sorry for not noticing them in time. I will start to work on them immediately.

**Comment by TerryGuo** [comment]
Hi @vladimirradosavljevic , I created another 3 PRs to address your comments:

1. Add shouldFormOverflowOp but without using the MathUsed #608 
2. Add shouldFormOverflowOp with using MathUsed #609 
3. Add shouldFormOverflowOp with using MathUsed, remove the constraint #610 

Looks to me the number changes very little. And there is a regress in #610 , do you want me to look into this?

**Comment by vladimirradosavljevic** [comment]
> Hi @vladimirradosavljevic , I created another 3 PRs to address your comments:
> 
> 1. Add shouldFormOverflowOp but without using the MathUsed [[EraVM] Enable shouldFormOverflowOp without MathUsed #608](https://github.com/matter-labs/era-compiler-llvm/pull/608)
> 2. Add shouldFormOverflowOp with using MathUsed [[EraVM] Enable shouldFormOverflowOp with MathUsed #609](https://github.com/matter-labs/era-compiler-llvm/pull/609)
> 3. Add shouldFormOverflowOp with using MathUsed, remove the constraint [[EraVM] Remove constraint for foldMultiplicationOverflowCheck #610](https://github.com/matter-labs/era-compiler-llvm/pull/610)
> 
> Looks to me the number changes very little. And there is a regress in #610 , do you want me to look into this?

@TerryGuo Thanks for looking into this, I really appreciate it!
Indeed, since there are no expected improvements, I don't think it is worth it to investigate further.
Regarding #610, optimization that you re enabled can generate `smul_with_overflow` and since we are not lowering `SMUL_LOHI` we are hitting an assert, so no need to look into this.

Thanks again for trying this!

**Comment by akiramenai** [comment]
Changed the commit message to make it better comprehensible without knowing the context.

**Review by akiramenai — APPROVED** [review]

**Pull request merged**